Composer/packages/server/src/controllers/publisher.ts

```diff
@@ -8,45 +8,76 @@ import merge from 'lodash/merge';
 import { pluginLoader, PluginLoader } from '../services/pluginLoader';
 import { BotProjectService } from '../services/project';
 import { runtimeFolder } from '../settings/env';
+
 const defaultPublishConfig = {
   name: 'default',
   type: 'localpublish',
+  configuration: JSON.stringify({}),
 };
 const DEFAULT_RUNTIME = 'CSharp';
 export const PublishController = {
   getTypes: async (req, res) => {
-    res.json(Object.keys(pluginLoader.extensions.publish));
+    res.json(
+      Object.keys(pluginLoader.extensions.publish)
+        .filter(i => pluginLoader.extensions.publish[i].plugin.name !== defaultPublishConfig.type)
+        .map(i => {
+          return {
+            name: pluginLoader.extensions.publish[i].plugin.name,
+            description: pluginLoader.extensions.publish[i].plugin.description,
+            features: {
+              history: pluginLoader.extensions.publish[i].methods.history ? true : false,
+              publish: pluginLoader.extensions.publish[i].methods.publish ? true : false,
+              status: pluginLoader.extensions.publish[i].methods.getStatus ? true : false,
+              rollback: pluginLoader.extensions.publish[i].methods.rollback ? true : false,
+            },
+          };
+        })
+    );
   },
   publish: async (req, res) => {
     const target = req.params.target;
     const user = await PluginLoader.getUserFromRequest(req);
-    const sensitiveSetting = req.body;
+    const { metadata, sensitiveSettings } = req.body;
     const projectId = req.params.projectId;
     const currentProject = await BotProjectService.getProjectById(projectId, user);
 
-    // find publish config by name.
-    const configs = currentProject.settings?.publishTargets?.filter(t => t.name === target) || [defaultPublishConfig];
-    const config = configs.length ? configs[0] : undefined;
-    const method = config ? config.type : undefined;
+    // deal with publishTargets not exist in settings
+    const publishTargets = currentProject.settings?.publishTargets || [];
```

Comment:
we could also provide `publishTargets` as a default in the settings object when a bot is created? Won't the local publish target come out of the box?

Reply:
Yes, I think we can add it, but in order to prevend the older bot, I think we still need to add this check.

Reply:
the local publish target is handled by Composer and does not have to exist in the settings.
That said, yes, this field _should_ exist in the default settings shape.

Reply:
I add the publishTargets as a field in default settings object.